// phyloviz_graph/index.js
var VivaGraph = require('vivagraphjs');
var graphFunc = require('./graphFunctions.js');

function constructGraph(graph, canvasID, callback){

  var graphObject = {

          graphInput: graph, //phyloviz input format
          graphGL: VivaGraph.Graph.graph(), //graph format used by Vivagraph
          nodeColor: 0x009ee8, //default nodeColor for the application
          DefaultnodeSize: 25, //default node size for the application
          datasetID: null, //id of the dataset
          width: null, //canvas width
          height: null, //canvas height
          prevSplitTreeValue: 0, //prev value used in splitTree method
          prevNLVvalue: 0, //prev value used in NLV graph method
          addedLinks: {}, //links added when we use NLV graph
          removedLinks: {}, //links removed when we use SplitTree
          container: canvasID

    }


  var graphFunctions = graphFunc();  //Functions to be applied to the graphObject object. graphFunctions.js

  graphFunctions.init(graphObject);
  graphFunctions.initLayout(graphObject);
  graphFunctions.initGraphics(graphObject);
  graphFunctions.initRenderer(graphObject);
  //graphFunctions.generateDOMLabels(graphObject);
  //graphFunctions.adjustScale(graphObject);

  callback(graphObject);
   
}

module.exports = constructGraph;
